### ISSUE TITLE
TypeError: Cannot read property 'apply' of undefined [reqOptions.callback.apply(self, m.params) in manager.js]

### ISSUE THREAD

**Issue by scottdavis99** [pofider/node-script-manager#1]
Hi, 

My name is Scott Davis. We are using JSReport in our application, and noticed a show-stopper bug was introduced in node-script-manager on roughly 2015-04-09. This bug was introduced without a corresponding SemVer bump in the version number. (In other words, last week script-manager#0.2.0 was working in production; this week script-manager#0.2.0 fails with this bug.)

The error thrown is this:

```
/Users/scott/code/local/hdslabs/doorways/node_modules/
jsreport/node_modules/script-manager/lib/manager.js:77

                reqOptions.callback.apply(self, m.params);

                                   ^

TypeError: Cannot read property 'apply' of undefined
    at ChildProcess.<anonymous> (/Users/scott/code/local/hdslabs/doorways/node_modules/
jsreport/node_modules/script-manager/lib/manager.js:77:36)
    at ChildProcess.emit (events.js:129:20)
    at handleMessage (child_process.js:324:10)
    at Pipe.channel.onread (child_process.js:352:11)

[nodemon] app crashed - waiting for file changes before starting...
```

Doing a bit of console logging, I discovered that the function call should be `reqOptions.cb`, not `reqObject.callback`.

```
m:
{ action: 'callback',
  rid: '636b7d90-e2be-11e4-b367-c77209aa2568',
  pid: 84187,
  params: [ null ] }

reqOptions:
{ execModulePath: '/Users/scott/code/local/hdslabs/doorways/node_modules/jsreport/lib/render/renderExecution.js',
  timeout: undefined,
  rid: '636b7d90-e2be-11e4-b367-c77209aa2568',
  isDone: false,
  cb: [Function] }
```

Here is the patched code. It is a single line -- the last `reqObject.cb.apply` call. (I can also send you a pull request if you'd prefer).

```
self.workersCluster.on("message", function (m) {
    if (m.action === "callback") {
        var reqOptions = _.findWhere(self._runningRequests, { rid: m.rid });
        console.log('m:');
        console.dir(m);
        console.log('reqOptions:');
        console.dir(reqOptions);
        m.params.push(function(){
            var args = Array.prototype.slice.call(arguments);
            if (args.length && args[0]) {
                args[0] = args[0].message;
            }
            self.workersCluster.send({
                action: "callback-response",
                rid: m.rid,
                params: args
            })
        });
        reqOptions.cb.apply(self, m.params);
    }
```

**Comment by pofider** [comment]
Hi Scott,
thanks for the bug report, but I think the problem is somewhere else.

I have introduced the new breaking change in script-manager 0.2 which requires following signature of all scripts:
```js
module.exports = function(inputs, callback, done) {
...
}
```

Where in script-manager 0.1 scripts' signature was

```js
module.exports = function(inputs, done) {
...
}
```

Unfortunately script-manager was wrongly referenced in the jsreport 0.3 and this breaking change started to bug it. jsreport 0.3 is using the old srcipts signature and undesirably calling new callback feature which throws the exception you describe. So the problem of your code is probably wrong signature of the script.

The reference for script-manager was fixed in 0.4 so hopefully we don't get into such a situation again. I am sorry for the problems it caused.

**Comment by scottdavis99** [comment]
Hi Jan,

Thanks for your feedback. We’re obviously big fans of your work here at HDS Labs. I work with Scott Fredericksen (cc’d), who you’ve been exchanging emails with on another thread.

We rev’ed to jsreports#0.4.0, and that seemed to fix the issue. 

Hopefully you understand our frustration that last week’s deployment to production (using script-manager#0.2.0) worked perfectly, yet this week’s deployment to production failed with exactly the same version of script-manager#0.2.0. Changing method signatures without rev’ing the SemVer made this bug quite tricky to track down, especially since it was in an underlying transitive dependency. 

This situation could’ve been avoided completely if:
1. You referenced the dependency on script-manager in jsreports’s package.json by explicit version number — 0.2.0 — instead of *.
2. You rev’ed the version of script-manager from 0.2.0 to 0.3.0 when you changed the method signature. 

That way jsreport#0.3.0 would’ve been able to pull down the appropriate dependency, and upgrading to jsreport#0.4.0 would’ve upgraded its dependencies accordingly. 

Hopefully you don’t feel that I’m lecturing you here — believe me, we’ve all made mistakes like this before, including me! My only goal with this is to turn it into a learning experience so that we don’t get caught by the same issue again in the future. We want to continue using your awesome software for a long time.

Thanks for all of the effort you’ve put into jsreports, including the website, documentation, and examples. 

Cheers,
s
Scott Davis
scott@davisworld.org

**Comment by pofider** [comment]
Thanks for very valuable advises. I really appreciate it.

1. jsreport is since 0.4 using "~" to reference script-manager and I will change it to match just exact version. I understand the frustration when the same version of jsreport suddenly starts to behave differently

2. Just to make it clear here. In npm there are versions 0.1.0, 0.1.1 and 0.2.0 of script-manager, where the signature is changed in 0.2.0. That is fine for npm, but unfortunately I didn't make tag in git for 0.1.0 and 0.1.1 what caused  part of the problems. 

To summarize: 
- I will let jsreport to reference dependencies only with explicit versions
- Every release will be properly tagged in git 

Thanks again, if anything else pops up what could be improved, please let me know.

**Comment by scottdavis99** [comment]
Jan,

You rock!

Thx,
s
Scott Davis
scott@davisworld.org
